op-e2e/e2eutils/l2oo/helper.go

```diff
@@ -0,0 +1,130 @@
+package l2oo
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"math/big"
+	"testing"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-bindings/bindings"
+	"github.com/ethereum-optimism/optimism/op-chain-ops/genesis"
+	"github.com/ethereum-optimism/optimism/op-e2e/e2eutils/wait"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/stretchr/testify/require"
+)
+
+type L2OOHelper struct {
+	t       *testing.T
+	require *require.Assertions
+	client  *ethclient.Client
+	l2oo    *bindings.L2OutputOracle
+
+	// Nil when read-only
+	transactOpts *bind.TransactOpts
+	rollupCfg    *rollup.Config
+}
```

Comment:
Just for my own understanding is it normal to set some properties nil sometimes in golang? My intuition here is you would need to make L200Helper and l2HelperReadOnly seperate structs repeating t, require, client, and l2oo props between the 2. I never would have thought of making pointers nullable

Reply:
So since they're non or unexported fields (the first letter is not upper cased), they cannot be accessed outside the golang package. As such, it is intended to not reference the internal variables of the `L2OOHelper` struct directly, but rather to use the public-facing, exposed api accessor and mutator methods. If these are implemented correctly, the internal behavior of making internal fields nullable is abstracted away from external users and IMO it's acceptable behavior. I think monadic type representations are just better all around, though.